### ISSUE TITLE
Fragments on Interfaces

### ISSUE THREAD

**Issue by guymers** [ghostdogpr/caliban#617]
I am attempting to do something similar to the union types example here https://graphql.org/learn/schema/#union-types

```scala
@GQLInterface
sealed trait Character {
  def id: String
  def name: String
}
object Character {
  case class Human(
    id: String,
    name: String,
    height: Int
  ) extends Character

  case class Droid(
    id: String,
    name: String,
    primaryFunction: String
  ) extends Character
}

case class Starship(
  id: String,
  name: String,
  length: Float,
)

// union SearchResult = Human | Droid | Starship
type SearchResult = Either[Character, Starship]

implicit val schemaSearchResult: Schema[Any, SearchResult] = {
  eitherUnionSchema("SearchResult")
}

def eitherUnionSchema[RL, RR, L, R](name: String)(implicit
  evL: Schema[RL, L],
  evR: Schema[RR, R],
): Schema[RL with RR, Either[L, R]] = new Schema[RL with RR, Either[L, R]] {

  override def toType(isInput: Boolean, isSubscription: Boolean): __Type = {
    val typeL = evL.toType_(isInput = isInput, isSubscription = isSubscription)
    val typeR = evR.toType_(isInput = isInput, isSubscription = isSubscription)

    Types.makeUnion(
      name = Some(name),
      description = None,
      subTypes = typeR :: typeL.possibleTypes.getOrElse(Nil),
    )
  }

  override def resolve(value: Either[L, R]): Step[RL with RR] = {
    value match {
      case Left(value) => evL.resolve(value)
      case Right(value) => evR.resolve(value)
    }
  }
}

case class SearchArgs(text: String)
case class Query(
  search: SearchArgs => List[SearchResult]
)
```

The above results in the schema I expect, but it is missing `interface Character`:

```graphql
schema {
  query: Query
}

union SearchResult = Starship | Droid | Human

type Droid implements Character {
  id: String!
  name: String!
  primaryFunction: String!
}

type Human implements Character {
  id: String!
  name: String!
  totalCredits: Int!
}

type Query {
  search(text: String!): [SearchResult!]!
}

type Starship {
  id: String!
  name: String!
  length: Float!
}
```

Adding a `dummy: Option[Character]` to the query to force `interface Character` to be defined allows the query in the example to be run:
```graphql
{
  search(text: "a") {
    __typename
    ... on Character {
      name
    }
    ... on Human {
      height
    }
    ... on Droid {
      primaryFunction
    }
    ... on Starship {
      name
      length
    }
  }
}
```

But it does not return any fields defined in the `Character` fragment
```json
{
  "search": [
    {
      "__typename": "Human",
      "height": 100
    },
    {
      "__typename": "Droid",
      "primaryFunction": "z"
    },
    {
      "__typename": "Starship",
      "name": "a",
      "length": 1000.0
    }
  ]
}
```

**Comment by ghostdogpr** [comment]
Looks like a bug. Thanks for the reproducer, I will have a look some time this week.

**Comment by ghostdogpr** [comment]
@guymers there was actually 2 different bugs. I fixed it in https://github.com/ghostdogpr/caliban/pull/619. Any chance you can try my branch and confirm it works for you? I added a test case for your example but maybe you have a more complex case.

**Comment by guymers** [comment]
Can confirm that #619 fixes my case, cheers

**Comment by ghostdogpr** [comment]
Thanks! I merged the PR, you can use snapshot `0.9.2+32-2be9b0a7-SNAPSHOT` until the next release.
